feature_creation_scripts/arctic_ocean_regions/arctic_ocean_regions.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
+
+import shapely
+import numpy
+import xarray
+import os
+import matplotlib.pyplot as plt
+import pyproj
+import gzip
+import shutil
+import tempfile
+
+from geometric_features import GeometricFeatures, FeatureCollection
+from geometric_features.feature_collection import _round_coords
+
+from geometric_features.download import download_files
+from geometric_features.utils import write_feature_names_and_tags
+
+
+def get_longest_contour(contourValue, author):
+
+    def stereo_to_lon_lat(x, y):
+        return pyproj.transform(inProj, outProj, x, y)
+
+    bathymetryFileName = 'IBCAO_V3_500m_SM.grd.gz'
+    if not os.path.exists(bathymetryFileName):
+        print('Downloading IBCAO bathymetry data...')
+        # download
+        baseURL = 'https://www.ngdc.noaa.gov/mgg/bathymetry/arctic/grids/version3_0'
+
+        download_files([bathymetryFileName], baseURL, './')
+
+    print('Decompressing and reading IBCAO bathymetry data...')
+    infile = gzip.open(bathymetryFileName, 'rb')
+    tmp = tempfile.NamedTemporaryFile(delete=False)
+    shutil.copyfileobj(infile, tmp)
+    infile.close()
+    tmp.close()
+    ds = xarray.open_dataset(tmp.name)
+    os.unlink(tmp.name)
+
+    # plot contours
+    plt.figure()
+    cs = plt.contour(ds.x.values, ds.y.values, ds.z, (contourValue,))
+    paths = cs.collections[0].get_paths()
+
+    pathLengths = [len(paths[i]) for i in range(len(paths))]
+    iLongest = numpy.argmax(pathLengths)
+
+    p = paths[iLongest]
+    v = p.vertices
+    x = v[:, 0]
+    y = v[:, 1]
+
+    # Arctic stereographic
+    inProj = pyproj.Proj(init='epsg:3995')
+    # lon/lat
+    outProj = pyproj.Proj(init='epsg:4326')
+    lon, lat = pyproj.transform(inProj, outProj, x, y)
+
+    poly = shapely.geometry.Polygon([(i[0], i[1]) for i in zip(x, y)])
+
+    epsilon = 1e-14
+    minY = numpy.amin(y)
+    wedge = shapely.geometry.Polygon([(epsilon, minY),
+                                      (epsilon**2, -epsilon),
+                                      (0, epsilon),
+                                      (-epsilon**2, -epsilon),
+                                      (-epsilon, minY),
+                                      (epsilon, minY)])
+
+    difference = poly.difference(wedge)
```

Comment:
here is where I get an error, possibly because this code was originally written by @xylar to work with `bedmap` in the Southern Ocean. The error message is:
```
TopologyException: Input geom 0 is invalid: Self-intersection at or near point -1114116.2216271362 -2232860.7958771652 at -1114116.2216271362 -2232860.7958771652
Traceback (most recent call last):
  File "arctic_ocean_regions.py", line 224, in <module>
    main()
  File "arctic_ocean_regions.py", line 196, in main
    fcContour300 = get_longest_contour(contourValue=-300., author=author)
  File "arctic_ocean_regions.py", line 83, in get_longest_contour
    difference = poly.difference(wedge)
  File "/lcrc/group/acme/milena/miniconda3.7/envs/conda_geometric_features/lib/python3.7/site-packages/shapely/geometry/base.py", line 616, in difference
    return geom_factory(self.impl['difference'](self, other))
  File "/lcrc/group/acme/milena/miniconda3.7/envs/conda_geometric_features/lib/python3.7/site-packages/shapely/topology.py", line 70, in __call__
    self._check_topology(err, this, other)
  File "/lcrc/group/acme/milena/miniconda3.7/envs/conda_geometric_features/lib/python3.7/site-packages/shapely/topology.py", line 38, in _check_topology
    self.fn.__name__, repr(geom)))
shapely.errors.TopologicalError: The operation 'GEOSDifference_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.polygon.Polygon object at 0x2b28d45a0be0>
```

To try to debug, I plotted the result of the `cs` contour, and the two variables `poly` and `wedge`, which are all attached below, respectively:
![cs_contour](https://user-images.githubusercontent.com/8237579/67893676-a6f15280-fb1c-11e9-83bd-f520f45c333e.png)
![poly](https://user-images.githubusercontent.com/8237579/67893684-ac4e9d00-fb1c-11e9-9d56-da733b80c2b5.png)
![wedge](https://user-images.githubusercontent.com/8237579/67893688-ae186080-fb1c-11e9-875f-0507576d00b3.png)

so it seems to me that there is no intersection between `wedge` and `poly`, but I  am not sure how to fix it..

Reply:
I see 2 problems.  The main one is that your contour hit s the edge of the boundary, unlike the one I was using.  You need a closed contour so you're going to have to fill in the bit that's getting connected with a straight line with two edges that follow the edge of you polar domain.  A self-intersecting shape is invalid according to the `shapely` library.

The much more minor problem is that your wedge is at 0 degrees longitude, not +/-180.  So it's going to divide your shape in a way we don't want.  Just make the y values positive instead of negative and you should be good.

Reply:
One way to fix the contour might be to put topography values of zero all around the edges of your topography file on the polar grid.  That way, a -300 would also follow the edges of your plotting domain.

Reply:
Thanks @xylar: all your suggestions worked.